### PR TITLE
Add file content and owner modification probes

### PR DESCRIPTION
**Pull request by mmat11**
```
File              Program                              Verdict  Duration (us)  Insns  States  Peak states
----------------  -----------------------------------  -------  -------------  -----  ------  -----------
EventProbe.bpf.o  fentry__commit_creds                 success           1899    740      35           35
EventProbe.bpf.o  fentry__do_renameat2                 success            218     68       4            4
EventProbe.bpf.o  fentry__do_unlinkat                  success            155     50       2            2
EventProbe.bpf.o  fentry__mnt_want_write               success            133     37       3            3
EventProbe.bpf.o  fentry__taskstats_exit               success          86209  26453    1397           78
EventProbe.bpf.o  fentry__tcp_close                    success           1358    474      26           26
EventProbe.bpf.o  fentry__tty_write                    success           1585    561      25           25
EventProbe.bpf.o  fentry__vfs_rename                   success         203489  79651    3119          405
EventProbe.bpf.o  fentry__vfs_unlink                   success            174     37       3            3
EventProbe.bpf.o  fexit__chmod_common                  success         102471  39551    1559          242
EventProbe.bpf.o  fexit__chown_common                  success         105521  39551    1559          242
EventProbe.bpf.o  fexit__do_filp_open                  success         107216  40449    1573          244
EventProbe.bpf.o  fexit__do_truncate                   success         105723  39583    1561          244
EventProbe.bpf.o  fexit__inet_csk_accept               success           1173    419      25           25
EventProbe.bpf.o  fexit__tcp_v4_connect                success           1214    422      25           25
EventProbe.bpf.o  fexit__tcp_v6_connect                success           1196    422      25           25
EventProbe.bpf.o  fexit__vfs_rename                    success           3423   1309      42           42
EventProbe.bpf.o  fexit__vfs_unlink                    success         106861  40420    1571          243
EventProbe.bpf.o  fexit__vfs_write                     success         104344  39562    1560          243
EventProbe.bpf.o  kprobe__chmod_common                 success            128     43       1            1
EventProbe.bpf.o  kprobe__chown_common                 success            119     41       1            1
EventProbe.bpf.o  kprobe__commit_creds                 success           1922    740      35           35
EventProbe.bpf.o  kprobe__do_renameat2                 success            208     68       4            4
EventProbe.bpf.o  kprobe__do_truncate                  success            192     63       3            3
EventProbe.bpf.o  kprobe__do_unlinkat                  success            158     50       2            2
EventProbe.bpf.o  kprobe__mnt_want_write               success            132     37       3            3
EventProbe.bpf.o  kprobe__taskstats_exit               success          86878  26453    1397           78
EventProbe.bpf.o  kprobe__tcp_close                    success           1430    474      26           26
EventProbe.bpf.o  kprobe__tcp_v4_connect               success            154     50       2            2
EventProbe.bpf.o  kprobe__tcp_v6_connect               success            145     50       2            2
EventProbe.bpf.o  kprobe__tty_write                    success           1565    561      25           25
EventProbe.bpf.o  kprobe__vfs_rename                   success         210256  79648    3120          406
EventProbe.bpf.o  kprobe__vfs_unlink                   success            133     39       4            4
EventProbe.bpf.o  kprobe__vfs_write                    success            282     54       1            1
EventProbe.bpf.o  kretprobe__chmod_common              success         105099  39561    1560          243
EventProbe.bpf.o  kretprobe__chown_common              success         106245  39561    1560          243
EventProbe.bpf.o  kretprobe__do_filp_open              success         108032  40449    1573          244
EventProbe.bpf.o  kretprobe__do_truncate               success         105759  39561    1560          243
EventProbe.bpf.o  kretprobe__inet_csk_accept           success           1214    419      25           25
EventProbe.bpf.o  kretprobe__tcp_v4_connect            success           1188    432      26           26
EventProbe.bpf.o  kretprobe__tcp_v6_connect            success           1189    432      26           26
EventProbe.bpf.o  kretprobe__vfs_rename                success           3378   1298      41           41
EventProbe.bpf.o  kretprobe__vfs_unlink                success         106097  40409    1570          242
EventProbe.bpf.o  kretprobe__vfs_write                 success         107164  39561    1560          243
EventProbe.bpf.o  sched_process_exec                   success         200834  67486    2987          292
EventProbe.bpf.o  sched_process_fork                   success          83060  26868    1416           99
EventProbe.bpf.o  tracepoint_syscalls_sys_exit_setsid  success            704    262      14           14
----------------  -----------------------------------  -------  -------------  -----  ------  -----------
Done. Processed 1 files, 0 programs. Skipped 47 files, 0 programs.
```

### CODE REVIEW

**Review by pkoutsovasilis — COMMENTED** [review]
don't we need `vfs_writev` as well?

**Comment by mmat11** [comment]
> don't we need vfs_writev as well?

yes, I think so, good catch!

there are also some kernels failing in CI, for the kprobe version, I'll fix that

File: GPL/Events/File/Probe.bpf.c [hunk 1]
```diff
@@ -532,7 +543,151 @@ int BPF_KRETPROBE(kretprobe__chmod_common, int ret)
...
+    struct ebpf_events_state state = {};
+
+    struct file *filp;
+    if (FUNC_ARG_READ_PTREGS(filp, do_truncate, filp)) {
+        bpf_printk("kprobe__do_truncate: error reading filp\n");
+        return 0;
+    }
+
+    struct path path    = BPF_CORE_READ(filp, f_path);
+    state.truncate.path = &path;
```
Comment (nicholasberlin) [inline GPL/Events/File/Probe.bpf.c:564]: Seems like this is the address of the stack variable and accessing it later through `ebpf_events_state__get(EBPF_EVENTS_STATE_TRUNCATE);` will cause problems.

Am I missing something?

File: GPL/Events/File/Probe.bpf.c [hunk 2]
```diff
@@ -532,7 +543,151 @@ int BPF_KRETPROBE(kretprobe__chmod_common, int ret)
...
+}
+
+SEC("kretprobe/do_truncate")
+int BPF_KRETPROBE(kretprobe__do_truncate, int ret)
+{
+    struct ebpf_events_state *state = ebpf_events_state__get(EBPF_EVENTS_STATE_TRUNCATE);
+    if (!state)
+        goto out;
+
+    do_truncate__exit(state->truncate.path, ret);
```
Comment (nicholasberlin) [inline GPL/Events/File/Probe.bpf.c:601]:
```suggestion
    (void)do_truncate__exit(state->truncate.path, ret);
```
explicitly ignore the return value? perhaps, `do_truncate__exit` should return void?

File: GPL/Events/File/Probe.bpf.c [hunk 3]
```diff
@@ -532,7 +543,151 @@ int BPF_KRETPROBE(kretprobe__chmod_common, int ret)
...
+}
+
+SEC("kprobe/vfs_write")
+int BPF_KPROBE(kprobe__vfs_write, struct file *file, const char *buf, size_t count, loff_t *pos)
+{
+    struct ebpf_events_state state = {};
+
+    struct path path = BPF_CORE_READ(file, f_path);
+    state.write.path = &path;
+    ebpf_events_state__set(EBPF_EVENTS_STATE_WRITE, &state);
```
Comment (nicholasberlin) [inline GPL/Events/File/Probe.bpf.c:614]: Same question here about storing what seems like a stack address.

File: GPL/Events/File/Probe.bpf.c [hunk 4]
```diff
@@ -532,7 +543,151 @@ int BPF_KRETPROBE(kretprobe__chmod_common, int ret)
...
+
+    return 0;
+}
+
+static int vfs_write__exit(struct path *path, int ret)
+{
+    if (ret)
```
Comment (nicholasberlin) [inline GPL/Events/File/Probe.bpf.c:621]: I thought `vfs_write` returned the number of bytes written. Is that wrong?

File: GPL/Events/File/Probe.bpf.c [hunk 5]
```diff
@@ -532,7 +543,151 @@ int BPF_KRETPROBE(kretprobe__chmod_common, int ret)
...
+}
+
+SEC("kretprobe/vfs_write")
+int BPF_KRETPROBE(kretprobe__vfs_write, int ret)
+{
+    struct ebpf_events_state *state = ebpf_events_state__get(EBPF_EVENTS_STATE_WRITE);
+    if (!state)
+        goto out;
+
+    vfs_write__exit(state->write.path, ret);
```
Comment (nicholasberlin) [inline GPL/Events/File/Probe.bpf.c:648]:
```suggestion
    (void)vfs_write__exit(state->write.path, ret);
```
explicitly ignore the return? make `vfs_write__exit` a void function?

File: GPL/Events/File/Probe.bpf.c (hunk 4 above)
Comment (mmat11) [reply GPL/Events/File/Probe.bpf.c:621]: yeah should be right

File: GPL/Events/File/Probe.bpf.c (hunk 1 above)
Comment (mmat11) [reply GPL/Events/File/Probe.bpf.c:564]: should be right

File: testing/testrunner/tests.go [hunk 6]
```diff
@@ -390,32 +390,55 @@ func TestFileDeleteContainer(et *EventsTraceInstance) {
...
 	if err := json.Unmarshal(outputStr, &binOutput); err != nil {
 		TestFail("failed to unmarshal json", err)
 	}
 
-	var fileModifyEvent FileModifyEvent
+	eventsCount := 4 // chmod, write, writev, truncate
+	events := make([]FileModifyEvent, 0, eventsCount)
 	for {
+		var event FileModifyEvent
+		fmt.Println("GREPME", len(events), events)
```
Comment (pkoutsovasilis) [inline testing/testrunner/tests.go:408]: do we really need this?

File: non-GPL/Events/EventsTrace/EventsTrace.c [hunk 7]
```diff
@@ -856,6 +861,10 @@ static void out_network_connection_closed_event(struct ebpf_net_event *evt)
 
 static int event_ctx_callback(struct ebpf_event_header *evt_hdr)
 {
+    if (g_print_features_init)
+        while (!g_features_printed)
```
Comment (pkoutsovasilis) [inline non-GPL/Events/EventsTrace/EventsTrace.c:865]: just to be sure here; this "polling" is just for cosmetic reasons to make sure that the main will print first all the feature messages and then print any captured ebpf events, right @mmat11 ?

Comment (mmat11) [reply non-GPL/Events/EventsTrace/EventsTrace.c:865]: yes, but tests are still failing, I am still debugging this

Comment (mmat11) [reply non-GPL/Events/EventsTrace/EventsTrace.c:865]: I can remove it, but doesn't hurt to have an explicit wait step

**Review by Tacklebox — APPROVED** [review]
LGTM :shipit:

**Pull request merged**